docs/admin/reference/settings.md
# Settings

There is one required setting, although specific plugins may have additional required
settings.

- `SECRET_KEY <secret-key-setting>`

!!! note
    For more information on how to specify settings see the [`Applying Settings`](#).

Pulp uses three types of settings:

- `Django settings` Pulp is configuring
- `Redis settings` Pulp is using
- `Pulp defined settings`

## Django Settings

Below is a list of the most common Django settings Pulp users typically use. Pulp is a Django
project, so any [Django setting](https://docs.djangoproject.com/en/4.2/ref/settings/) can be set.

### SECRET_KEY

In order to get a pulp server up and running a [Django SECRET_KEY](https://docs.djangoproject.com/en/4.2/ref/settings/#secret-key) *must* be
provided.
The following code snippet can be used to generate a random SECRET_KEY.

```python linenums="1"
import secrets

chars = 'abcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*(-_=+)'
print(''.join(secrets.choice(chars) for i in range(50)))
```

### DB_ENCRYPTION_KEY

The file location of a symmetric fernet key that Pulp uses to encrypt sensitive fields in the
database. Default location is `/etc/pulp/certs/database_fields.symmetric.key`.
The key is automatically generated by default with the pulp-oci-images. The key can be generated
independently but it must be a url-safe base64-encoded string of 32 random bytes.

To generate a key with openssl:

```
openssl rand -base64 32 > /etc/pulp/certs/database_fields.symmetric.key
```

This file can contain multiple such keys (one per line). The key in the first line will be used
for encryption but all others will still be attempted to decrypt old tokens. This can help you to
rotate this key in the following way:

1. Shut down all Pulp services (api, content and worker processes).
1. Add a new key at the top of the key file.
1. Call `pulpcore-manager rotate-db-key`.
1. Remove the old key (on the second line) from the key file.
1. Start the Pulp services again.

For a zero downtime key rotation you can follow the slightly more complex recipe:

1. Add a new key at the bottom of the key file.
1. Restart the Pulp services in the usual phased manner.
1. Swap the keys in the key file.
1. Restart the Pulp services again.
1. Call `pulpcore-manager rotate-db-key`.
1. Remove the old key (on the second line) from the key file.
1. Restart the Pulp services for the last time.

### DATABASES

By default Pulp uses PostgreSQL on localhost. PostgreSQL is the only supported database. For
instructions on how to configure the database, refer to `database installation <database-install>`.

### DEFAULT_FILE_STORAGE

By default, Pulp uses the local filesystem to store files. The default option which
uses the local filesystem is `pulpcore.app.models.storage.FileSystem`.

For more information about different Pulp storage options, see the
`storage documentation `.

### REDIRECT_TO_OBJECT_STORAGE

When set to `True` access to artifacts is redirected to the corresponding Cloud storage
configured in `DEFAULT_FILE_STORAGE` using pre-authenticated URLs. When set to `False`
artifacts are always served by the content app instead.

Defaults to `True`; ignored for local file storage.

### MEDIA_ROOT

The location where Pulp will store files. By default this is `/var/lib/pulp/media`.

This only affects storage location when `DEFAULT_FILE_STORAGE` is set to
`pulpcore.app.models.storage.FileSystem`. See the `storage documentation ` for
more info.

It should have permissions of:

- mode: 750
- owner: pulp (the account that pulp runs under)
- group: pulp (the group of the account that pulp runs under)
- SELinux context: system_u:object_r:pulpcore_var_lib_t:s0

### LOGGING

By default Pulp logs at an INFO level to syslog. For all possible configurations please
refer to [Django documenation on logging](https://docs.djangoproject.com/en/4.2/topics/logging/#configuring-logging)

Enabling DEBUG logging is a common troubleshooting step. See the `enabling-debug-logging`
documentation for details on how to do that.

### AUTHENTICATION_BACKENDS

By default, Pulp has two types of authentication enabled, and they fall back for each other:

1. Basic Auth which is checked against an internal users database
2. Webserver authentication that relies on the webserver to perform the authentication.

To change the authentication types Pulp will use, modify the `AUTHENTICATION_BACKENDS`
settings. See the [Django authentication documentation](https://docs.djangoproject.com/en/4.2/topics/auth/customizing/#authentication-backends) for more information.



## Redis Settings

!!! warning
    To enable usage of Redis the [CACHE_ENABLED] option must be set to `True`.


The following Redis settings can be set in your Pulp config:

- REDIS_URL
- REDIS_HOST
- REDIS_PORT
- REDIS_DB
- REDIS_PASSWORD

Below are some common settings used for Redis configuration.

### REDIS_HOST

The hostname for Redis.

### REDIS_PORT

The port for Redis.

### REDIS_PASSWORD

The password for Redis.



## Pulp Settings

Pulp defines the following settings itself:



### API_ROOT

A string containing the path prefix for the Pulp API. This is used by the REST API when forming
URLs to refer clients to other parts of the REST API and by the `pulpcore-api` application to
match incoming URLs. Pulp appends the string `api/v3/` to this setting.

Defaults to `/pulp/`. After the application appends `api/v3/` it makes the V3 API by default
serve from `/pulp/api/v3/`.

### WORKING_DIRECTORY

The directory used by workers to stage files temporarily. This defaults to
`/var/lib/pulp/tmp/`.

It should have permissions of:

- mode: 750
- owner: pulp (the account that pulp runs under)
- group: pulp (the group of the account that pulp runs under)
- SELinux context: system_u:object_r:pulpcore_var_lib_t:s0

!!! note
    It is recommended that `WORKING_DIRECTORY` and `MEDIA_ROOT` exist on the same storage
    volume for performance reasons. Files are commonly staged in the `WORKING_DIRECTORY` and
    validated before being moved to their permanent home in `MEDIA_ROOT`.


### CHUNKED_UPLOAD_DIR

A relative path inside the DEPLOY_ROOT directory used exclusively for uploaded chunks. The
uploaded chunks are stored in the default storage specified by `DEFAULT_FILE_STORAGE`. This
option allows users to customize the actual place where chunked uploads should be stored within
the declared storage. The default, `upload`, is sufficient for most use cases. A change to
this setting only applies to uploads created after the change.



### CONTENT_ORIGIN

A string containing the protocol, fqdn, and port where the content app is reachable by users.
This is used by `pulpcore` and various plugins when referring users to the content app.
For example if the API should refer users to content at using http to pulp.example.com on port
24816, (the content default port), you would set: `https://pulp.example.com:24816`. The default is "".
When set to "" (empty string), the `base_url` for Distributions is a relative path.
This means the API returns relative URLs without the protocol, fqdn, and port.


### HIDE_GUARDED_DISTRIBUTIONS

If activated, the distributions that are protected by a content guard will not be shown on the
directory listing in the content app. Defaults to `False`.



### CONTENT_PATH_PREFIX

A string containing the path prefix for the content app. This is used by the REST API when
forming URLs to refer clients to the content serving app, and by the content serving application
to match incoming URLs.

Defaults to `/pulp/content/`.



### CONTENT_APP_TTL

The number of seconds before a content app should be considered lost.

Defaults to `30` seconds.



### CACHE_ENABLED

Store cached responses from the content app into Redis. This setting improves the performance
of the content app under heavy load for similar requests. Defaults to `False`.

!!! note
    The entire response is not stored in the cache. Only the location of the file needed to
    recreate the response is stored. This reduces database queries and allows for many
    responses to be stored inside the cache.


### CACHE_SETTINGS

Dictionary with tunable settings for the cache:

- `EXPIRES_TTL` - Number of seconds entries should stay in the cache before expiring.

Defaults to `600` seconds.

!!! note
    Set to `None` to have entries not expire.
    Content app responses are always invalidated when the backing distribution is updated.


### DOMAIN_ENABLED

!!! note
    This feature is provided as a tech-preview


Enable the `Domains feature to enable multi-tenancy capabilities <domains>`. All installed
plugins must be Domain compatible for Pulp to start. Defaults to `False`.



### WORKER_TTL

The number of seconds before a worker should be considered lost.

Defaults to `30` seconds.



### REMOTE_USER_ENVIRON_NAME

The name of the WSGI environment variable to read for `webserver authentication
<webserver-authentication>`.

!!! warning
    Configuring this has serious security implications. See the [Django warning at the end of this
    section in their docs](https://docs.djangoproject.com/en/4.2/howto/auth-remote-user/#configuration) for more details.

Defaults to `'REMOTE_USER'`.



### ALLOWED_IMPORT_PATHS

One or more real filesystem paths that Remotes with filesystem paths can import from. For example
to allow a remote url of `file:///mnt/foo/bar/another/folder/` you could specify:

```
ALLOWED_IMPORT_PATHS = ['/mnt/foo/bar']  # only a subpath is needed
```

Defaults to `[]`, meaning `file:///` urls are not allowed in any Remote.



### ALLOWED_EXPORT_PATHS

One or more real filesystem paths that Exporters can export to. For example to allow a path of
`/mnt/foo/bar/another/folder/` you could specify:

```
ALLOWED_EXPORT_PATHS = ['/mnt/foo/bar']  # only a subpath is needed
```

Defaults to `[]` which means no path is allowed.



### ALLOWED_CONTENT_CHECKSUMS

!!! warning
    Enforcement of this setting in `pulpcore` and various plugins is not fully in place. It is
    possible that checksums not in this list may still be used in various places. This banner will
    be removed when it is believed all `pulpcore` and plugin code fully enforces this setting.


The list of content-checksums this pulp-instance is **allowed to use**. By default the following
are used:

```
ALLOWED_CONTENT_CHECKSUMS = ["sha224", "sha256", "sha384", "sha512"]
```

The entire set of supported checksums are: `md5`, `sha1`, `sha224`, `sha256`,
`sha384`, and `sha512`.

!!! warning
    Due to its use as the primary content-identifier, "sha256" **IS REQUIRED**. Pulp will
    fail to start if `"sha256"` is not found in this set.

Pulp can prohibit or allow checksums by setting the ALLOWED_CONTENT_CHECKSUMS setting.
Changing this setting requires a few steps.

First, before you change the setting, see how your Pulp instance will be impacted by this change by running:

`pulpcore-manager handle-artifact-checksums --report --checksums sha256,512`

Adjust `--checksums` as comma separated list of checksums types to match your needs.

!!! note
    If you already changed `ALLOWED_CONTENT_CHECKSUMS` in pulp settings you can leave out `--checksums`,
    and the checksums will be parsed from Pulp settings.

Before switching, any on-demand repos containing forbidden checksum digests needs to be synced with
`policy=immediate` to populate missing allowed checksums. This can heavily impact your disk space.
Alternatively, users can remove these offending repo versions followed by orphan cleanup.

If you have artifacts that do not conform to your ALLOWED_CONTENT_CHECKSUMS setting, they need to be re-hashed.
You can update them using:

`pulpcore-manager handle-artifact-checksums`

!!! warning
    If Pulp fails to start because forbidden checkums have been identified or required ones are
    missing, run `pulpcore-manager handle-artifact-checksums` command.

### DJANGO_GUID

Pulp uses `django-guid` to append correlation IDs to logging messages. Correlation IDs are
autogenerated by default but can also be sent as a header with each request. They are also
returned as a header in the response and are recorded in the `logging_cid` field of tasks.

For more information on how to configure the `DJANGO_GUID` setting, see the [django-guid
settings documentation](https://django-guid.readthedocs.io/en/latest/settings.html).



### ORPHAN_PROTECTION_TIME

The time specified in minutes for how long Pulp will hold orphan Content and Artifacts before
they become candidates for deletion by an orphan cleanup task. This should ideally be longer
than your longest running task otherwise any content created during that task could be cleaned
up before the task finishes. Default is 1440 minutes (24 hours).





### UPLOAD_PROTECTION_TIME, TMPFILE_PROTECTION_TIME and TASK_PROTECTION_TIME

Pulp uses `tasks`, `uploads` and `pulp temporary files` to pass data from the api to worker tasks.
These options allow to specify a timeinterval in minutes used for cleaning up stale entries.
If set to 0, automatic cleanup is disabled, which is the default.



### TASK_DIAGNOSTICS

The default setting is `False`. When set to `True`, each task records various diagnostics (listed below)
and stores them as separate artifacts. To download the data, issue GET requests to `${TASK_HREF}profile_artifacts/`.
The artifacts are cleaned up automatically by the orphan cleanup.

- memory.datum - the task's max resident set size in MB
- pyinstrument.html - the output of the pyinstrument profiler, if installed



### ANALYTICS

If `True`, Pulp will anonymously post analytics information to
[https://analytics.pulpproject.org/](https://analytics.pulpproject.org/) and aids in project decision making. See the
`analytics docs ` for more info on exactly what is posted along with an example.

Defaults to `True`.

## Kafka Settings

!!! note
    Kafka integration functionality is in tech preview and may change based on user feedback.

See [librdkafka configuration documentation](https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md)
for details on client configuration properties.

### KAFKA_BOOTSTRAP_SERVERS

`bootstrap.servers` value for the client. Specifies endpoint(s) for the kafka client. Kafka integration is disabled if
unspecified.

### KAFKA_SECURITY_PROTOCOL


pulpcore/app/settings.py
"""
Django settings for the Pulp Platform application

Never import this module directly, instead `from django.conf import settings`, see
<https://docs.djangoproject.com/en/3.2/topics/settings/#using-settings-in-python-code>

For the full list of settings and their values, see
<https://docs.djangoproject.com/en/3.2/ref/settings/>
"""

import sys

from contextlib import suppress
from importlib import import_module
from logging import getLogger
from pathlib import Path

from cryptography.fernet import Fernet
from django.core.exceptions import ImproperlyConfigured
from django.db import connection

from pulpcore import constants


try:
    import sentry_sdk

    sentry_sdk.init()
except ImportError:
    pass

if sys.version_info < (3, 10):
    # Python 3.9 has a rather different interface for `entry_points`.
    # Let's use a compatibility version.
    from importlib_metadata import entry_points
else:
    from importlib.metadata import entry_points

# Build paths inside the project like this: BASE_DIR / ...
BASE_DIR = Path(__file__).absolute().parent

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/3.2/howto/deployment/checklist/

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = False

ALLOWED_HOSTS = ["*"]

DEPLOY_ROOT = Path("/var/lib/pulp")
MEDIA_ROOT = str(DEPLOY_ROOT / "media")  # Django 3.1 adds support for pathlib.Path

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/3.2/howto/static-files/

STATIC_URL = "/assets/"
STATIC_ROOT = DEPLOY_ROOT / STATIC_URL.strip("/")

DEFAULT_FILE_STORAGE = "pulpcore.app.models.storage.FileSystem"
REDIRECT_TO_OBJECT_STORAGE = True

WORKING_DIRECTORY = DEPLOY_ROOT / "tmp"
FILE_UPLOAD_TEMP_DIR = WORKING_DIRECTORY

CHUNKED_UPLOAD_DIR = "upload"

# List of upload handler classes to be applied in order.
FILE_UPLOAD_HANDLERS = ("pulpcore.app.files.HashingFileUploadHandler",)

# SECURITY WARNING: this should be set to a unique, unpredictable value
SECRET_KEY = "SECRET"

# Key used to encrypt fields in the database
DB_ENCRYPTION_KEY = "/etc/pulp/certs/database_fields.symmetric.key"

# API Root
API_ROOT = "/pulp/"
API_ROOT_REWRITE_HEADER = None

# Application definition

INSTALLED_APPS = [
    # django stuff
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "django.contrib.postgres",
    "import_export",
    # third-party
    "django_filters",
    "django_guid",
    "drf_spectacular",
    "rest_framework",
    # pulp core app
    "pulpcore.app",
]

# Enumerate the installed Pulp plugins during the loading process for use in the status API
INSTALLED_PULP_PLUGINS = []

for entry_point in entry_points(group="pulpcore.plugin"):
    plugin_app_config = entry_point.load()
    INSTALLED_PULP_PLUGINS.append(entry_point.name)
    INSTALLED_APPS.append(plugin_app_config)

# Optional apps that help with development, or augment Pulp in some non-critical way
OPTIONAL_APPS = [
    "crispy_forms",
    "django_extensions",
    "storages",
]

for app in OPTIONAL_APPS:
    # only import if app is installed
    with suppress(ImportError):
        import_module(app)
        INSTALLED_APPS.append(app)

MIDDLEWARE = [
    "django_guid.middleware.guid_middleware",
    "django.middleware.security.SecurityMiddleware",
    "whitenoise.middleware.WhiteNoiseMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
    "pulpcore.middleware.DomainMiddleware",
    "pulpcore.middleware.APIRootRewriteMiddleware",
]

AUTHENTICATION_BACKENDS = [
    "django.contrib.auth.backends.ModelBackend",
    "pulpcore.backends.ObjectRolePermissionBackend",
]

ROOT_URLCONF = "pulpcore.app.urls"

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [BASE_DIR / "templates"],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ],
        },
    },
]

WSGI_APPLICATION = "pulpcore.app.wsgi.application"

REST_FRAMEWORK = {
    "URL_FIELD_NAME": "pulp_href",
    "DEFAULT_FILTER_BACKENDS": ("pulpcore.filters.PulpFilterBackend",),
    "DEFAULT_PAGINATION_CLASS": "rest_framework.pagination.LimitOffsetPagination",
    "PAGE_SIZE": 100,
    "DEFAULT_PERMISSION_CLASSES": ("pulpcore.app.access_policy.AccessPolicyFromDB",),
    "DEFAULT_AUTHENTICATION_CLASSES": (
        "rest_framework.authentication.BasicAuthentication",
        "rest_framework.authentication.SessionAuthentication",
    ),
    "UPLOADED_FILES_USE_URL": False,
    "DEFAULT_VERSIONING_CLASS": "rest_framework.versioning.URLPathVersioning",
    "DEFAULT_SCHEMA_CLASS": "pulpcore.openapi.PulpAutoSchema",
}

# Password validation
# https://docs.djangoproject.com/en/3.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {"NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator"},
    {"NAME": "django.contrib.auth.password_validation.MinimumLengthValidator"},
    {"NAME": "django.contrib.auth.password_validation.CommonPasswordValidator"},
    {"NAME": "django.contrib.auth.password_validation.NumericPasswordValidator"},
]


# Internationalization
# https://docs.djangoproject.com/en/3.2/topics/i18n/

LANGUAGE_CODE = "en-us"

TIME_ZONE = "UTC"

USE_I18N = "USE_I18N", True

USE_L10N = True

USE_TZ = True


# A set of default settings to use if the configuration file in
# /etc/pulp/ is missing or if it does not have values for every setting

# https://docs.djangoproject.com/en/3.2/ref/settings/#databases
DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.postgresql",
        "NAME": "pulp",
        "USER": "pulp",
        "CONN_MAX_AGE": 0,
    },
}

# Redis default config
REDIS_URL = None
REDIS_HOST = None
REDIS_PORT = None
REDIS_DB = 0
REDIS_PASSWORD = None
REDIS_SSL = False
REDIS_SSL_CA_CERTS = None

# Social_django settings
SOCIAL_AUTH_URL_NAMESPACE = "social"
SOCIAL_AUTH_POSTGRES_JSONFIELD = True
SOCIAL_AUTH_JSONFIELD_CUSTOM = "django.db.models.JSONField"

# https://docs.djangoproject.com/en/3.2/ref/settings/#logging and
# https://docs.python.org/3/library/logging.config.html
LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "formatters": {
        "simple": {"format": "pulp [%(correlation_id)s]: %(name)s:%(levelname)s: %(message)s"}
    },
    "filters": {"correlation_id": {"()": "django_guid.log_filters.CorrelationId"}},
    "handlers": {
        "console": {
            "class": "logging.StreamHandler",
            "formatter": "simple",
            "filters": ["correlation_id"],
        }
    },
    "loggers": {
        "": {
            # The root logger
            "handlers": ["console"],
            "level": "INFO",
            "filters": ["correlation_id"],
        },
        "django_guid": {
            "handlers": ["console"],
            "level": "WARNING",
            "propagate": False,
        },
    },
}

# Custom access policies indexed by the url_pattern of the view.
# This will only take effect when combined with the
# `pulpcore.app.access_policy.AccessPoliciesFromSettings` permission class.
ACCESS_POLICIES = {}

DRF_ACCESS_POLICY = {"reusable_conditions": ["pulpcore.app.global_access_conditions"]}

CONTENT_ORIGIN = ""
CONTENT_PATH_PREFIX = "/pulp/content/"

API_APP_TTL = 120  # The heartbeat is called from gunicorn notify (defaulting to 45 sec).
CONTENT_APP_TTL = 30
WORKER_TTL = 30

# how long to protect ephemeral items in minutes
ORPHAN_PROTECTION_TIME = 24 * 60

# Custom cleaup intervals
# for the following, if set to 0, the corresponding cleanup task is disabled
UPLOAD_PROTECTION_TIME = 0
TASK_PROTECTION_TIME = 0
TMPFILE_PROTECTION_TIME = 0

REMOTE_USER_ENVIRON_NAME = "REMOTE_USER"

AUTHENTICATION_JSON_HEADER = ""
AUTHENTICATION_JSON_HEADER_JQ_FILTER = ""
AUTHENTICATION_JSON_HEADER_OPENAPI_SECURITY_SCHEME = {}

ALLOWED_IMPORT_PATHS = []

ALLOWED_EXPORT_PATHS = []

# https://docs.pulpproject.org/pulpcore/configuration/settings.html#pulp-cache
CACHE_ENABLED = False
CACHE_SETTINGS = {
    "EXPIRES_TTL": 600,  # 10 minutes
}

# The time a RemoteArtifact will be ignored after failure.
# In on-demand, if a fetching content from a remote failed due to corrupt data,
# the corresponding RemoteArtifact will be ignored for that time (seconds).
REMOTE_CONTENT_FETCH_FAILURE_COOLDOWN = 5 * 60  # 5 minutes

SPECTACULAR_SETTINGS = {
    "SERVE_URLCONF": ROOT_URLCONF,
    "DEFAULT_GENERATOR_CLASS": "pulpcore.openapi.PulpSchemaGenerator",
    "DEFAULT_SCHEMA_CLASS": "pulpcore.openapi.PulpAutoSchema",
    "ENUM_ADD_EXPLICIT_BLANK_NULL_CHOICE": False,
    "COMPONENT_SPLIT_REQUEST": True,
    "COMPONENT_NO_READ_ONLY_REQUIRED": True,
    "GENERIC_ADDITIONAL_PROPERTIES": None,
    "DISABLE_ERRORS_AND_WARNINGS": not DEBUG,
    "POSTPROCESSING_HOOKS": [
        "drf_spectacular.hooks.postprocess_schema_enums",
        "pulpcore.openapi.hooks.add_info_hook",
    ],
    "TITLE": "Pulp 3 API",
    "DESCRIPTION": "Fetch, Upload, Organize, and Distribute Software Packages",
    "VERSION": "v3",
    "CONTACT": {
        "name": "Pulp Team",
        "email": "pulp-list@redhat.com",
        "url": "https://pulpproject.org",
    },
    "LICENSE": {
        "name": "GPLv2+",
        "url": "https://raw.githubusercontent.com/pulp/pulpcore/master/LICENSE",
    },
}

# What kinds of checksums is this pulp-instance _allowed to use_ ?
# NOTE : "sha256"" IS REQUIRED - Pulp will fail to start if it is not found in this set
# NOTE: specifying checksums that are not listed under ALL_KNOWN_CONTENT_CHECKSUMS will fail
#       at startup
ALLOWED_CONTENT_CHECKSUMS = ["sha224", "sha256", "sha384", "sha512"]

DEFAULT_AUTO_FIELD = "django.db.models.AutoField"

TASK_DIAGNOSTICS = False

ANALYTICS = True

HIDE_GUARDED_DISTRIBUTIONS = False

DOMAIN_ENABLED = False

SHELL_PLUS_IMPORTS = [
    "from pulpcore.app.util import get_domain, get_domain_pk, set_domain, get_url, extract_pk",
    "from pulpcore.tasking.tasks import dispatch, cancel_task, wakeup_worker",
]

# What percentage of available-workers will pulpimport use at a time, max
# By default, use all available workers.
IMPORT_WORKERS_PERCENT = 100

# Kafka settings
KAFKA_BOOTSTRAP_SERVERS = None  # kafka integration disabled by default
KAFKA_TASKS_STATUS_TOPIC = "pulpcore.tasking.status"
KAFKA_TASKS_STATUS_PRODUCER_SYNC_ENABLED = False
KAFKA_PRODUCER_POLL_TIMEOUT = 0.1
KAFKA_SECURITY_PROTOCOL = "plaintext"
KAFKA_SSL_CA_PEM = None
KAFKA_SASL_MECHANISM = None
KAFKA_SASL_USERNAME = None
KAFKA_SASL_PASSWORD = None

# opentelemetry settings
OTEL_ENABLED = False

# HERE STARTS DYNACONF EXTENSION LOAD (Keep at the very bottom of settings.py)
# Read more at https://www.dynaconf.com/django/
from dynaconf import DjangoDynaconf, Validator  # noqa

# Validators
storage_validator = (
    Validator("REDIRECT_TO_OBJECT_STORAGE", eq=False)
    | Validator("DEFAULT_FILE_STORAGE", eq="pulpcore.app.models.storage.FileSystem")
    | Validator("DEFAULT_FILE_STORAGE", eq="storages.backends.azure_storage.AzureStorage")
    | Validator("DEFAULT_FILE_STORAGE", eq="storages.backends.s3boto3.S3Boto3Storage")
    | Validator("DEFAULT_FILE_STORAGE", eq="storages.backends.gcloud.GoogleCloudStorage")
)
storage_validator.messages["combined"] = (
    "'REDIRECT_TO_OBJECT_STORAGE=True' is only supported with the local file, S3, GCP or Azure"
    "storage backend configured in DEFAULT_FILE_STORAGE."
)

cache_enabled_validator = Validator("CACHE_ENABLED", eq=True)
redis_url_validator = Validator("REDIS_URL", must_exist=True, when=cache_enabled_validator)
redis_host_validator = Validator("REDIS_HOST", must_exist=True, when=cache_enabled_validator)
redis_port_validator = Validator("REDIS_PORT", must_exist=True, when=cache_enabled_validator)
cache_validator = redis_url_validator | (redis_host_validator & redis_port_validator)
cache_validator.messages["combined"] = (
    "CACHE_ENABLED is enabled but it requires to have REDIS configured. Please check "
    "https://docs.pulpproject.org/pulpcore/configuration/settings.html#redis-settings "
    "for more information."
)

sha256_validator = Validator(
    "ALLOWED_CONTENT_CHECKSUMS",
    cont="sha256",
    messages={
        "operations": "ALLOWED_CONTENT_CHECKSUMS MUST contain 'sha256' - Pulp's "
        "content addressable storage relies on sha256 to identify entities."
    },
)

unknown_algs_validator = Validator(
    "ALLOWED_CONTENT_CHECKSUMS",
    condition=lambda x: len(set(x).difference(constants.ALL_KNOWN_CONTENT_CHECKSUMS)) == 0,
    messages={
        "condition": (
            "ALLOWED_CONTENT_CHECKSUMS may only contain algorithms known to pulp - see "
            "constants.ALL_KNOWN_CONTENT_CHECKSUMS for the allowed list."
        )
    },
)

api_root_validator = Validator(
    "API_ROOT",
    condition=lambda x: x.startswith("/") and x.endswith("/"),
    messages={
        "condition": ("The API_ROOT must start and end with a '/', currently it is '{value}'")
    },
)

json_header_auth_class_restframework_validator = Validator(
    "REST_FRAMEWORK__DEFAULT_AUTHENTICATION_CLASSES",
    cont="pulpcore.app.authentication.JSONHeaderRemoteAuthentication",
)

authentication_json_header_validator = Validator(
    "AUTHENTICATION_JSON_HEADER",
    startswith="HTTP_",
    must_exist=True,
    when=json_header_auth_class_restframework_validator,
    messages={"startswith": 'The AUTHENTICATION_JSON_HEADER must start with "HTTP_"'},
)

authentication_json_header_jq_filter_validator = Validator(
    "AUTHENTICATION_JSON_HEADER_JQ_FILTER",
    startswith=".",
    must_exist=True,
    when=json_header_auth_class_restframework_validator,
    messages={"startswith": 'The AUTHENTICATION_JSON_HEADER_JQ_FILTER must start with "."'},
)

json_header_auth_validator = (
    authentication_json_header_validator & authentication_json_header_jq_filter_validator
)

authentication_json_header_openapi_security_scheme_setting_validator = Validator(
    "AUTHENTICATION_JSON_HEADER_OPENAPI_SECURITY_SCHEME", len_min=1
)
authentication_json_header_openapi_security_scheme_validator = Validator(
    "AUTHENTICATION_JSON_HEADER_OPENAPI_SECURITY_SCHEME",
    when=authentication_json_header_openapi_security_scheme_setting_validator,
    is_type_of=dict,
    messages={"is_type_of": "{name} must be a dictionary."},
)


def otel_middleware_hook(settings):
    data = {"dynaconf_merge": True}
    if settings.OTEL_ENABLED:
        data["MIDDLEWARE"] = ["pulpcore.middleware.DjangoMetricsMiddleware"]
    return data


settings = DjangoDynaconf(
    __name__,
    ENVVAR_PREFIX_FOR_DYNACONF="PULP",
    ENV_SWITCHER_FOR_DYNACONF="PULP_ENV",
    PRELOAD_FOR_DYNACONF=[
        "{}.app.settings".format(plugin_name) for plugin_name in INSTALLED_PULP_PLUGINS
    ],
    ENVVAR_FOR_DYNACONF="PULP_SETTINGS",
    load_dotenv=False,
    validators=[
        api_root_validator,
        cache_validator,
        sha256_validator,
        storage_validator,
        unknown_algs_validator,
        json_header_auth_validator,
        authentication_json_header_openapi_security_scheme_validator,
    ],
    post_hooks=otel_middleware_hook,
)
# HERE ENDS DYNACONF EXTENSION LOAD (No more code below this line)

_logger = getLogger(__name__)


if not (
    Path(sys.argv[0]).name in ["pytest", "sphinx-build"]
    or (len(sys.argv) >= 2 and sys.argv[1] in ["collectstatic", "openapi"])
):
    try:
        with open(DB_ENCRYPTION_KEY, "rb") as key_file:
            Fernet(key_file.read())
    except Exception as ex:
        raise ImproperlyConfigured(
            ("Could not load DB_ENCRYPTION_KEY file '{file}': {err}").format(
                file=DB_ENCRYPTION_KEY, err=ex
            )
        )


FORBIDDEN_CHECKSUMS = set(constants.ALL_KNOWN_CONTENT_CHECKSUMS).difference(
    ALLOWED_CONTENT_CHECKSUMS
)

_SKIPPED_COMMANDS_FOR_CONTENT_CHECKS = [
    "handle-artifact-checksums",
    "migrate",
    "collectstatic",
    "openapi",
]

if not (len(sys.argv) >= 2 and sys.argv[1] in _SKIPPED_COMMANDS_FOR_CONTENT_CHECKS):
    try:
        with connection.cursor() as cursor:
            for checksum in ALLOWED_CONTENT_CHECKSUMS:
                # can't import Artifact here so use a direct db connection
                cursor.execute(f"SELECT count(pulp_id) FROM core_artifact WHERE {checksum} IS NULL")
                row = cursor.fetchone()
                if row[0] > 0:
                    raise ImproperlyConfigured(
                        (
                            "There have been identified artifacts missing checksum '{}'. "
                            "Run 'pulpcore-manager handle-artifact-checksums' first to populate "
                            "missing artifact checksums."
                        ).format(checksum)
                    )
            for checksum in FORBIDDEN_CHECKSUMS:
                # can't import Artifact here so use a direct db connection
                cursor.execute(
                    f"SELECT count(pulp_id) FROM core_artifact WHERE {checksum} IS NOT NULL"
                )
                row = cursor.fetchone()
                if row[0] > 0:
                    raise ImproperlyConfigured(
                        (
                            "There have been identified artifacts with forbidden checksum '{}'. "
                            "Run 'pulpcore-manager handle-artifact-checksums' first to unset "
                            "forbidden checksums."
                        ).format(checksum)
                    )

            # warn if there are remote artifacts with checksums but no allowed checksums
            cond = " AND ".join([f"{c} IS NULL" for c in constants.ALL_KNOWN_CONTENT_CHECKSUMS])
            no_checksum_query = f"SELECT pulp_id FROM core_remoteartifact WHERE {cond}"
            cond = " AND ".join([f"{c} IS NULL" for c in ALLOWED_CONTENT_CHECKSUMS])
            cursor.execute(
                f"SELECT count(pulp_id) FROM core_remoteartifact WHERE {cond} AND "
                f"pulp_id NOT IN ({no_checksum_query})"
            )
            row = cursor.fetchone()
            if row[0] > 0:
                _logger.warning(
                    (
                        "Warning: detected remote content without allowed checksums. "
                        "Run 'pulpcore-manager handle-artifact-checksums --report' to "
                        "view this content."
                    )
                )

    except ImproperlyConfigured as e:
        raise e
    except Exception:
        # our check could fail if the table hasn't been created yet or we can't get a db connection
        pass
    finally:
        connection.close()

if settings.API_ROOT_REWRITE_HEADER:
    api_root = "/<path:api_root>/"
else:
    api_root = settings.API_ROOT
settings.set("V3_API_ROOT", api_root + "api/v3/")  # Not user configurable
settings.set("V3_DOMAIN_API_ROOT", api_root + "<slug:pulp_domain>/api/v3/")
settings.set("V3_API_ROOT_NO_FRONT_SLASH", settings.V3_API_ROOT.lstrip("/"))
settings.set("V3_DOMAIN_API_ROOT_NO_FRONT_SLASH", settings.V3_DOMAIN_API_ROOT.lstrip("/"))
